infinispan-embedded/src/main/java/org/hibernate/ogm/datastore/infinispan/InfinispanDialect.java

```diff
@@ -228,6 +247,104 @@ public void forEachTuple( ModelConsumer consumer, TupleTypeContext tupleTypeCont
 		return (KeyProvider<EK, AK, ISK>) provider.getKeyProvider();
 	}
 
+	@Override
+	public ClosableIterator<Tuple> callStoredProcedure( String storedProcedureName, ProcedureQueryParameters queryParameters, TupleContext tupleContext ) {
+		validate( queryParameters );
+		EmbeddedCacheManager embeddedCacheManager = getCacheManager().getCacheManager();
+		Cache<String, String> cache = embeddedCacheManager.getCache( "___stored_procedures", true );
+		String className = cache.getOrDefault( storedProcedureName, storedProcedureName );
+		Callable<?> callable = instantiate( storedProcedureName, className );
+		setParams( storedProcedureName, queryParameters, callable );
+		Object res = execute( storedProcedureName, embeddedCacheManager, callable );
+		return CollectionHelper.newClosableIterator( extractTuples( storedProcedureName, res ) );
+	}
+
+	private void validate(ProcedureQueryParameters queryParameters) {
+		List<Object> positionalParameters = queryParameters.getPositionalParameters();
+		if ( positionalParameters != null && positionalParameters.size() > 0 ) {
+			throw log.dialectDoesNotSupportPositionalParametersForStoredProcedures( getClass() );
+		}
+	}
+
+	private Object execute(String storedProcedureName, EmbeddedCacheManager embeddedCacheManager, Callable<?> callable) {
+		AtomicReference<Object> ref = new AtomicReference<>();
+		try {
+			return embeddedCacheManager.executor()
+					.submitConsumer( ecm -> execute( callable ), ( a, r, e ) -> {
+						if ( e != null ) {
+							if ( e instanceof InfinispanDialect.ExceptionWrapper ) {
+								throw log.cannotExecuteStoredProcedure( storedProcedureName, e.getCause() );
+							}
+							throw log.cannotExecuteStoredProcedure( storedProcedureName, e );
+						}
+						ref.set( r );
+					} )
+					.thenCompose( v -> CompletableFuture.supplyAsync( ref::get ) )
+					.get();
+		}
+		catch ( InterruptedException | ExecutionException e ) {
+			throw log.cannotExecuteStoredProcedure( storedProcedureName, e );
+		}
+	}
+
+	private static Callable<?> instantiate(String storedProcedureName, String className) {
+		try {
+			return (Callable<?>) ReflectionHelper.instantiate( className );
+		}
+		catch (ClassCastException | ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+			throw log.cannotInstantiateStoredProcedure( storedProcedureName, className, e );
+		}
+	}
+
+	private static void setParams( String storedProcedureName, ProcedureQueryParameters queryParameters, Callable<?> callable ) {
+		try {
+			ReflectionHelper.setFields( callable, queryParameters.getNamedParameters() );
```

Comment:
It would be very nice for people using it if the error being thrown could say exactly wich "setter" is expected, when a NoSuchMethodException is thrown. Could we improve the exception errors?